File: 2019/practice-round/Mural.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int t;          //number of test cases
    int n;          //number of sections
    int toBePaint;  //number of sections that will be painted
    int beauty;     //value of beauty score
    int max;        //maximum beauty
    char* wall;     //representation of wall's sections

    scanf("%d", &t);    //read the number of tests from stdin

    //repeat for each test
    for (int counter = 1; counter <= t; counter++) {

        scanf("%d", &n);            //read number of sections of the wall
        wall = (char*)malloc(n);    //allocate as much memory as needed
        scanf("%s", wall);          //read the wall's representation

        //the number of sections to be painted are the half in case of even number of sections, half plus one otherwise
        toBePaint = (n / 2) + (n % 2);
        beauty = 0;
        max = 0;
        
        for (int i = 0; i < n; i++) {
            beauty += (int) wall[i] - 48;   //increment the beauty score by the score of the i-th section
            if (i >= toBePaint)
                beauty -= (int) wall[i - toBePaint] - 48;   //decrease the beauty score by the score of the no-longer-be-painted section
            if (beauty > max)
                max = beauty;   //substitute the maximum beauty score if needed
        }

        printf("Case #%d: %d\n", counter, max);
        free(wall); //free memory for next test case
    }
}

File: 2019/practice-round/mural.py
import sys
from math import ceil

def total_beauty(N, b_values):

	# the number of paintable sectors is the upper integer half
	window_size = ceil(N / 2)

	# the paintable sectors are contiguous so we build a moving window over the wall
	window = b_values[0:window_size]
	max_beauty = sum(window)

	# each iteration the window moves of one sector to the right
	for i in range(window_size, len(b_values)):
		window.pop(0)
		window.append(b_values[i])

		# if the beauty is greater to the max, it becomes the new max
		temp_beauty = sum(window)
		if temp_beauty > max_beauty:
			max_beauty = temp_beauty

	return max_beauty

# read the numebr of test cases
T = int(sys.stdin.readline())
result = ''
for i in range(T):

	# read the values of the test case
	N = int(sys.stdin.readline())
	beauties = list(sys.stdin.readline().rstrip('\n'))
	b_values = [int(value) for value in beauties]
	
	# compute the maximum reachable beauty
	x = i + 1
	y = total_beauty(N, b_values)
	result += f'Case #{x}: {y}\n'

print(result)
